### PR TITLE
Keyboard shortcut prefix fix for Windows, Linux (Ctrl+Shift). Fixes #1069.

### LINKED ISSUE #1069: Ctrl+X not working in nano
- **OS version and name**: elementary OS Loki
- **Hyper.app version**: @https://github.com/zeit/hyper/commit/e31f975f788f3ced0175c5cffd203690d99c350c
- **Link of a [Gist](https://gist.github.com/) with the contents of your .hyper.js**: https://gist.github.com/WoodyWoodsta/722486e90139eba0138aa0df82140296
- **Relevent information from devtools** _(CMD+SHIFT+I on Mac OS, CTRL+SHIFT+I elsewhere)_: N/A
- **The issue is reproducible in vanilla Hyper.app**: Is Vanilla
## Issue

As of https://github.com/zeit/hyper/commit/be286c0d5aaa14528112213c9be3201384d2dcfa, <kbd>Ctrl</kbd>+<kbd>X</kbd> does not work in nano, since it is being caught by hyper for the cut command. I'm sure this may be one of many conflicting keyboard shortcuts. Perhaps a way to temporarily bypass the hyper accelerators would help for these cases.

### PR DESCRIPTION
**Pull request by nbolten**
This pull request fixes #1069 (and related issues like #1100).

### Overview

Interactive command line applications commonly bind to <kbd>Ctrl</kbd>+<kbd>&lt;key&gt;</kbd> keyboard shortcuts. For example, `nano` uses <kbd>Ctrl</kbd>+(<kbd>G</kbd>, <kbd>O</kbd>, <kbd>W</kbd>, <kbd>K</kbd>, <kbd>J</kbd>, <kbd>C</kbd>, <kbd>X</kbd>, <kbd>R</kbd>, <kbd>\\</kbd>, <kbd>U</kbd>, <kbd>T</kbd>, <kbd>_</kbd>, <kbd>V</kbd>) for essential functionality like reading and writing files and exiting the program. Other applications are similar, like `tmux` and `screen`, relying on keyboard shortcuts to manage the program's basic functionality.

This isn't an issue when using `hyper` on Macs, as the modifier prefix for interacting with the `hyper` windows is <kbd>Cmd</kbd>. On non-Macs (like Windows and Linux machines), the default prefix is `Ctrl`, which causes many conflicts.

This pull request proposes to address these conflicts by setting the default prefix on non-Macs to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>, which is common for terminal emulators in linux. The <kbd>Super</kbd> key has been suggested as an alternative, but this would conflict with several shortcuts built into the Windows OS as well as many linux-focused desktop environment like Unity and Gnome. To avoid conflicts with defaults that already used <kbd>Ctrl</kbd>+<kbd>Shift</kbd>, some shortcuts were changed to (hopefully) reasonable alternatives. This pull request does not change any shortcuts on Macs.

### Summary of changes

1. Default shortcut prefix on non-Macs changed from <kbd>Ctrl</kbd> to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>.

2. Some shortcuts that already used <kbd>Ctrl</kbd>+<kbd>Shift</kbd> were changed to be more similar to their Mac counterparts:
    - split vertically was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>E</kbd> to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd>

3. Some shortcuts already used <kbd>Ctrl</kbd>+<kbd>Shift</kbd> and are unchanged
    - copy is <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>C</kbd>
    - devtools is <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>I</kbd>
    - update plugins is <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>U</kbd>
    - `moveLeft2` and `moveRight2` are <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>{</kbd> and <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>}</kbd>

4. Some shortcuts were changed to (hopefully) reasonable alternatives
   - split horizontally was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>O</kbd> to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>E</kbd>
    - redo was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Z</kbd> (now undo) to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Y</kbd>
    - closing the entire window was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>W</kbd> (now close tab) to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Q</kbd>
    - full reload was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>R</kbd> (now reload) to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>F5</kbd>

5. I didn't know what to do with some of the shortcuts, so in those cases I replaced <kbd>Shift</kbd> with Z (please review/change!)
    - previous pane was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Alt</kbd>+<kbd>Tab</kbd> to <kbd>Ctrl</kbd>+<kbd>Z</kbd>+<kbd>Alt</kbd>+<kbd>Tab</kbd>.
    - `moveLeft1` was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Left</kbd> to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Z</kbd>+<kbd>Left</kbd>
    - `moveRight1` was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Right</kbd> to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Z</kbd>+<kbd>Right</kbd>
    - `moveLeft4` was changed from <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Tab</kbd> to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Z</kbd>+<kbd>Tab</kbd>

One of the shortcuts in that last case already involved five simultaneous keypresses, even on a Mac, so I'll humbly suggest it be switched to something else entirely!

I was unable to test whether the `undo` functionality was broken by using <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Z</kbd>+<kbd>&lt;key&gt;</kbd>. I'm not actually sure what `undo` is supposed to do in `hyper` (or the moveLeft1/2/3/4 options). Are they bound to any actions?

It may also be worthwhile to review Atom's shortcuts for linux + windows for better defaults. Windows version here https://github.com/atom/atom/blob/master/keymaps/win32.cson and linux version here https://github.com/atom/atom/blob/master/keymaps/linux.cson.

### From hyper's pull request template

- *To help whoever reviews your PR, it'd be extremely helpful for you to list whether your PR is ready to be merged*: 
  - This PR is ready to be merged
- *If there's anything left to do and if there are any related PRs*
  - Someone should review the new defaults to make sure they don't break conventions. I'm also unsure what to do with the 4 cases where I used <kbd>Z</kbd> instead of <kbd>Shift</kbd>, though it didn't seem to break anything for me.
  - Pull requests related to shortcuts would need to be updated, particularly #925 
- *It'd also be extremely helpful to enable us to update your PR incase we need to rebase or what-not by checking `Allow edits from maintainers`*
  - Done!

### CODE REVIEW

**Comment by yellowiscool** [comment]
I tested the diff on Hyper 1.0.0 Windows. The new shortcuts don't seem to work, Ctrl+Shift+T or Ctrl+Shift+A for example. Ctrl+A still selects the text and doesn't send the command.

It's on the good track though 😃

**Comment by nbolten** [comment]
Hi @yellowiscool! I did a build on Windows 10 (npm run dev + npm run pack + running the dist Hyper.exe) and Ctrl+Shift+T and Ctrl+Shift+W worked fine.

Is Ctrl+A supposed to select all the text in the current cursor? Neither that nor Ctrl+Shift+A work for me, nor does Menu>Edit>Select All.

This is running off of a downloaded zip copy of my fork.

**Comment by nbolten** [comment]
Ah, @yellowiscool, the `accelerators.js` file didn't exist in v1.0.0, so applying the diff won't do anything (the app won't even read `accelerators.js`). `accelerators.js` was added in commit be286c0d5aaa145.

**Comment by hahuang65** [comment]
I'd love for this to make it into a build sooner rather than later! I really want to use Hyper on Linux, but I need ctrl-a and ctrl-v to not be swallowed by the app and passed on to the shell.

**Comment by ags131** [comment]
Same here, I use ctrl+a in tmux and ctrl+w and ctrl+x in nano, this is currently a blocker for me using hyper until this is merged.

**Comment by celevra** [comment]
same here, this is bug not a feature, the fix has to be released asap. For me its also a show stopper

**Comment by hahuang65** [comment]
Any chance we can get a review on this? Would really love to use this...

**Comment by alanzanattadev** [comment]
An update on the current state of this issue ?

**Comment by internalfx** [comment]
Awwww....I assumed this had been merged when 1.1 came out.....back to running @nbolten fork again.

**Comment by sambanks** [comment]
I am also waiting on this one, would love to start using on my Linux desktop

**Comment by paulcbetts** [comment]
I'm :+1: on this idea, Ctrl-[LETTER] has too much going on to try to assign all of the shortcuts to it

File: app/accelerators.js [hunk 1]
```diff
@@ -11,38 +11,38 @@ const applicationMenu = { // app/menu.js
...
 
   // Plugins menu
-  updatePlugins: 'Shift+U',
+  updatePlugins: isMac ? 'Shift+U' : 'U',
 
   // Window menu
   minimize: 'M',
-  showPreviousTab: 'Alt+Left',
-  showNextTab: 'Alt+Right',
+  showPreviousTab: isMac ? 'Alt+Left' : '+Ctrl+PageDown',
```
Comment (paulcbetts) [inline app/accelerators.js]: Can this just be Ctrl-Shift-Tab?

File: app/accelerators.js [hunk 2]
```diff
@@ -11,38 +11,38 @@ const applicationMenu = { // app/menu.js
...
+  showNextTab: isMac ? 'Alt+Right' : '+Ctrl+PageUp',
```
Comment (paulcbetts) [inline app/accelerators.js]: Ctrl-Tab?

File: app/accelerators.js [hunk 3]
```diff
@@ -11,38 +11,38 @@ const applicationMenu = { // app/menu.js
...
   selectNextPane: 'Ctrl+Alt+Tab',
```
Comment (paulcbetts) [inline app/accelerators.js]: This shortcut doesn't work on Windows, it will invoke Alt-Tab

File: app/accelerators.js (hunk 1 above)
Comment (nbolten) [inline app/accelerators.js]: No strong opinions from me on this, but Ctrl+PageDown/Up is used by VS Code, Gnome Terminal, and Atom.

File: app/accelerators.js (hunk 3 above)
Comment (nbolten) [inline app/accelerators.js]: Ctrl+Alt+Tab/Ctrl+Shift+Alt+Tab also conflicts with the Accessibility Switcher on Gnome, that's no good...

iTerm uses Cmd+]/[.

tmux, screen, and Atom use special pane modes for pane manipulation, so e.g 'next pane' in Atom is Ctrl+k to enter pane mode, then Ctrl+n for 'next pane'.

I don't use panes and also don't have strong opinions on this one.

File: app/accelerators.js (hunk 1 above)
Comment (nbolten) [inline app/accelerators.js]: Actually, Ctrl+Tab and Ctrl+Shift+Tab is also used by Atom/Chrome, so maybe both could be used?

File: app/accelerators.js (hunk 3 above)
Comment (paulcbetts) [inline app/accelerators.js]: Maybe Ctrl-Shift-LeftArrow + Ctrl-Shift-RightArrow?

File: app/accelerators.js (hunk 1 above)
Comment (paulcbetts) [inline app/accelerators.js]: :+1:

**Comment by mienaikage** [comment]
Would it be possible to change the 'update plugins' shortcut? Ctrl+Shift+U is for inserting unicode characters in Ubuntu.

**Comment by ppot** [comment]
Some are missing. But can you update those key map with the desired keys in comments.
https://github.com/ppot/hyper/issues/11
```
{
 "devtools":"Cmd+Alt+I",
 "window:new":"Cmd+N",
 "window:minimize": "Cmd+M",
 "window:close":"Cmd+Shift+W",
 "tab:new":"Cmd+T",
 "tab:next":"Cmd+shift+]",
 "tab:prev":"Cmd+shift+[",
 "pane:next":"Cmd+]",
 "pane:prev":"Cmd+[",
 "pane:left":"Cmd+Alt+Left",
 "pane:right":"Cmd+Alt+Right",
 "pane:up":"Cmd+Alt+Up",
 "pane:down":"Cmd+Alt+Down",
 "pane:vertical":"Cmd+D",
 "pane:horizontal":"Cmd+Shift+D",
 "pane:close":"Cmd+W",
 "editor:undo":"Cmd+Z",
 "editor:redo":"Cmd+Shift+Z",
 "editor:cut":"Cmd+X",
 "editor:copy":"Cmd+C",
 "editor:paste":"Cmd+V",
 "editor:selectAll":"Cmd+A",
 "editor:clearBuffer":"Cmd+K"
}
```

**Comment by celevra** [comment]
why is this not in the actual release?
hyper is not usable under windows and linux without this pull request!
and no, ctrl+shift is no solution!

**Comment by none-da** [comment]
:+1:

**Comment by Stanzilla** [comment]
@ppot any chance to get this in any time soon?

**Comment by ppot** [comment]
@Stanzilla You mean my version? See https://github.com/zeit/hyper/pull/1509

Still need to make flexible Keymap before merging

**Comment by Stanzilla** [comment]
Oh, I see

**Comment by celevra** [comment]
three weaks gone, 2 new release and this isn't merged... again so much time lost where i cant use hyper, and i want to use hyper!

**Comment by ppot** [comment]
@celevra Look at Keymap PR

**Comment by RobertWHurst** [comment]
Would love to see this merged.

**Comment by apinter** [comment]
I just started to use Hyper a few hours ago on my Arch box and love it so much I put it on my F25 daily workstation as well, BUT the keybindings are making it... well not useless, but difficult to live with. 
Frankly I can't use vim and having difficulties with tmux as well.
Do you happened to have an estimate when can this fix go public?

File: app/accelerators.js (hunk 3 above)
Comment (ignatiusreza) [inline app/accelerators.js]: just throwing out some ideas.. in atom, the panes are indexed, and we can switch between them using `ctrl+[index]`.. would it be possible to have something similar?

**Comment by ignatiusreza** [comment]
any plan on making the shortcuts configurable?

File: app/accelerators.js (hunk 3 above)
Comment (chabou) [inline app/accelerators.js]: @ignatiusreza hyper-pane plugin has been made for this purpose: https://www.npmjs.com/package/hyper-pane

Comment (ignatiusreza) [inline app/accelerators.js]: wow! looking good! thanks @chabou :+1:

**Comment by ppot** [comment]
@ignatiusreza  Yes, #1509

**Comment by alexellis** [comment]
How can I fix the key-bindings to match with the Gnome Terminal? Is there somewhere I need to paste the {} json block quoted above?

**Comment by ppot** [comment]
@alexellis PR listed above Keymaps

**Comment by alexellis** [comment]
@ppot my question is how to apply / set those keycaps. It's not clear to me from the documentation / PR. Please could you explain?

**Comment by ppot** [comment]
@alexellis  Take a look at my PR Keymaps #1509

**Comment by nbolten** [comment]
Since a cooler keymaps implementation is on the way, I'm closing this PR. Hope #1509 gets reviewed/merged soon!

**Pull request closed without merging**